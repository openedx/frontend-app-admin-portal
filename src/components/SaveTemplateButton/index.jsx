import React from 'react';
import PropTypes from 'prop-types';
import { StatefulButton, Icon, Spinner } from '@edx/paragon';
import { CheckCircle } from '@edx/paragon/icons';
import { SubmissionError } from 'redux-form';

import { validateEmailTemplateFields } from '../../data/validation/email';
import { EMAIL_TEMPLATE_SOURCE_NEW_EMAIL, EMAIL_TEMPLATE_SUBJECT_KEY } from '../../data/constants/emailTemplate';
import SUBMIT_STATES from '../../data/constants/formSubmissions';
import { features } from '../../config';

export const SAVE_TEMPLATE_MODE = 'save';

class SaveTemplateButton extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      submitState: SUBMIT_STATES.DEFAULT,
    };

    this.timer = undefined;
    this.validateFormData = this.validateFormData.bind(this);
    this.handleSaveTemplate = this.handleSaveTemplate.bind(this);
    this.setButtonStateToDefault = this.setButtonStateToDefault.bind(this);
  }

  componentDidUpdate() {
    const { submitState } = this.state;
    if (submitState === SUBMIT_STATES.COMPLETE) {
      this.timer = setTimeout(() => {
        this.setButtonStateToDefault();
      }, 2000);
    }
  }

  handleSaveTemplate(formData) {
    const {
      setMode,
      saveTemplate,
      templateType,
      emailTemplateSource,
      emailTemplates,
    } = this.props;

    setMode('save');
    const newTemplateSource = emailTemplateSource === EMAIL_TEMPLATE_SOURCE_NEW_EMAIL;

    // Check the form validation in case user is saving new template
    // or there are no already saved templates.
    if (newTemplateSource || emailTemplates.allTemplates.length === 0) {
      this.validateFormData(formData);
    }

    const options = {
      email_type: templateType,
      email_subject: formData[EMAIL_TEMPLATE_SUBJECT_KEY],
      email_greeting: formData['email-template-greeting'],
      email_closing: formData['email-template-closing'],
      name: formData['template-name'] || formData['template-name-select'],
      ...(features.FILE_ATTACHMENT && {
        email_files: formData['email-template-files'],
      }),
    };

    if (!newTemplateSource) {
      options.id = emailTemplates[templateType]['template-id'];
    }

    this.setState({ submitState: SUBMIT_STATES.PENDING });
    return saveTemplate(options)
      .then(() => {
        this.setState({ submitState: SUBMIT_STATES.COMPLETE });
      })
      .catch((error) => {
        // backend to frontend field name map
        const fieldMap = {
          email_subject: 'email-template-subject',
          email_greeting: 'email-template-greeting',
          email_closing: 'email-template-closing',
          ...(features.FILE_ATTACHMENT && {
            email_files: 'email-template-files',
          }),
        };
        const { response, message } = error;
        const errors = {
          _error: [],
        };

        /* eslint-disable no-underscore-dangle */
        Object.entries(fieldMap).forEach(([key, value]) => {
          if (response && response.data && response.data[key]) {
            const msg = response.data[key][0];
            errors[value] = msg;
            errors._error.push(msg);
          }
        });

        // when no field specific error returned then show the main exception message
        if (Object.keys(errors).length === 1 && Object.keys(errors)[0] === '_error') {
          errors._error.push(message);
        }
        /* eslint-enable no-underscore-dangle */

        this.setState({ submitState: SUBMIT_STATES.DEFAULT });

        throw new SubmissionError(errors);
      });
  }

  setButtonStateToDefault() {
    this.setState({ submitState: SUBMIT_STATES.DEFAULT });
    clearTimeout(this.timer);
  }

  validateFormData(formData) {
    const emailTemplateNameKey = 'template-name';
    const errors = validateEmailTemplateFields(formData, emailTemplateNameKey);

    /* eslint-disable no-underscore-dangle */
    if (!formData[emailTemplateNameKey]) {
      const message = 'No template name provided. Please enter a template name.';
      errors[emailTemplateNameKey] = message;
      errors._error.unshift(message);
    }

    if (Object.keys(errors) > 1 || errors._error.length > 0) {
      throw new SubmissionError(errors);
    }
    /* eslint-enable no-underscore-dangle */
  }

  render() {
    const { submitState } = this.state;
    const { handleSubmit, disabled } = this.props;
    const buttonIconClasses = {
      default: 'outline-primary',
      pending: 'outline-primary',
      complete: 'outline-success',
    };

    return (
      <StatefulButton
        variant={buttonIconClasses[submitState]}
        className="save-template-btn"
        onClick={handleSubmit(this.handleSaveTemplate)}
        state={submitState}
        labels={{
          default: 'Save Template',
          pending: 'Saving Template...',
          complete: 'Template Saved',
        }}
        icons={{
          pending: <Spinner animation="border" variant="primary" size="sm" />,
          complete: <Icon src={CheckCircle} />,
        }}
        disabledStates={[SUBMIT_STATES.PENDING]}
        disabled={disabled}
      />
    );
  }
}

SaveTemplateButton.defaultProps = {
  emailTemplates: {
    allTemplates: [],
    assign: { 'template-id': 0 },
    remind: { 'template-id': 0 },
    revoke: { 'template-id': 0 },
  },
  disabled: false,
};

SaveTemplateButton.propTypes = {
  emailTemplateSource: PropTypes.string.isRequired,
  templateType: PropTypes.string.isRequired,
  saveTemplate: PropTypes.func.isRequired,
  setMode: PropTypes.func.isRequired,
  handleSubmit: PropTypes.func.isRequired,
  emailTemplates: PropTypes.shape(),
  disabled: PropTypes.bool,
};

export default SaveTemplateButton;
