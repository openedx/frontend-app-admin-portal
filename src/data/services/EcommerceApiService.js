import { getAuthenticatedHttpClient } from '@edx/frontend-platform/auth';

import { configuration } from '../../config';
import store from '../store';
import { EMAIL_TEMPLATE_SOURCE_FROM_TEMPLATE } from '../constants/emailTemplate';

class EcommerceApiService {
  static ecommerceBaseUrl = configuration.ECOMMERCE_BASE_URL;

  static apiClient = getAuthenticatedHttpClient

  static fetchCouponOrders(options) {
    const { enterpriseId } = store.getState().portalConfiguration;
    const queryParams = new URLSearchParams({
      page: 1,
      page_size: 50,
      filter: 'active',
      ...options,
    });
    const url = `${EcommerceApiService.ecommerceBaseUrl}/api/v2/enterprise/coupons/${enterpriseId}/overview/?${queryParams.toString()}`;
    return EcommerceApiService.apiClient().get(url);
  }

  static fetchCoupon(couponId) {
    const url = `${EcommerceApiService.ecommerceBaseUrl}/api/v2/enterprise/coupons/${couponId}/`;
    return EcommerceApiService.apiClient(
      {
        useCache: configuration.USE_API_CACHE,
      },
    ).get(url);
  }

  static fetchCouponDetails(couponId, options, { csv } = {}) {
    const endpoint = csv ? 'codes.csv' : 'codes';
    const queryParams = new URLSearchParams({
      page: 1,
      page_size: 50,
      ...options,
    });
    const url = `${EcommerceApiService.ecommerceBaseUrl}/api/v2/enterprise/coupons/${couponId}/${endpoint}/?${queryParams.toString()}`;
    return EcommerceApiService.apiClient().get(url);
  }

  static sendCodeAssignment(couponId, options) {
    const url = `${EcommerceApiService.ecommerceBaseUrl}/api/v2/enterprise/coupons/${couponId}/assign/`;
    return EcommerceApiService.apiClient().post(url, options);
  }

  static sendCodeReminder(couponId, options) {
    const url = `${EcommerceApiService.ecommerceBaseUrl}/api/v2/enterprise/coupons/${couponId}/remind/`;
    return EcommerceApiService.apiClient().post(url, options);
  }

  static sendCodeRevoke(couponId, options) {
    const url = `${EcommerceApiService.ecommerceBaseUrl}/api/v2/enterprise/coupons/${couponId}/revoke/`;
    return EcommerceApiService.apiClient().post(url, options);
  }

  static fetchCodeSearchResults(options) {
    const { enterpriseId } = store.getState().portalConfiguration;
    let url = `${EcommerceApiService.ecommerceBaseUrl}/api/v2/enterprise/coupons/${enterpriseId}/search/`;
    if (options) {
      const queryParams = new URLSearchParams(options);
      url += `?${queryParams.toString()}`;
    }
    return EcommerceApiService.apiClient().get(url);
  }

  static fetchEmailTemplate(options) {
    const { enterpriseId } = store.getState().portalConfiguration;
    let url = `${EcommerceApiService.ecommerceBaseUrl}/api/v2/enterprise/offer-assignment-email-template/${enterpriseId}/`;
    if (options) {
      const queryParams = new URLSearchParams(options);
      url += `?${queryParams.toString()}`;
    }
    return EcommerceApiService.apiClient().get(url);
  }

  static saveTemplate(options) {
    const { enterpriseId } = store.getState().portalConfiguration;
    const { emailTemplateSource } = store.getState().emailTemplate;
    let url = `${EcommerceApiService.ecommerceBaseUrl}/api/v2/enterprise/offer-assignment-email-template/${enterpriseId}/`;
    if (emailTemplateSource === EMAIL_TEMPLATE_SOURCE_FROM_TEMPLATE) {
      url = `${url}${options.id}/`;
      return EcommerceApiService.apiClient().put(url, options);
    }
    return EcommerceApiService.apiClient().post(url, options);
  }
}

export default EcommerceApiService;
