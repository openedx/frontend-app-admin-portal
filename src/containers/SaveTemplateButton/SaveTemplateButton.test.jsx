import React from 'react';
import renderer from 'react-test-renderer';
import { MemoryRouter } from 'react-router-dom';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { Provider } from 'react-redux';
import { fireEvent, render, screen } from '@testing-library/react';
import { SubmissionError } from 'redux-form';

import EcommerceApiService from '../../data/services/EcommerceApiService';
import {
  SAVE_TEMPLATE_REQUEST,
  EMAIL_TEMPLATE_FIELD_MAX_LIMIT,
  OFFER_ASSIGNMENT_EMAIL_SUBJECT_LIMIT,
  EMAIL_TEMPLATE_SUBJECT_KEY,
  EMAIL_TEMPLATE_SOURCE_NEW_EMAIL,
} from '../../data/constants/emailTemplate';
import SaveTemplateButton from './index';

jest.mock('../../data/services/EcommerceApiService');
const mockStore = configureMockStore([thunk]);
const initialState = {
  portalConfiguration: {
    enterpriseId: 'test-enterprise-id',
  },
  emailTemplate: {
    saving: false,
    allTemplates: [],
    assign: {
      'template-id': 1,
      'email-template-subject': 'email-template-subject',
      'email-template-greeting': 'email-template-greeting',
      'email-template-body': 'email-template-body',
      'email-template-files': 'email-template-files',
    },
  },
};
const store = mockStore({
  ...initialState,
});
const formData = {
  'template-name': 'Template from portal',
  'email-template-subject': 'Subject',
  'email-template-greeting': 'Greeting',
  'email-template-closing': 'Closing',
  'email-template-files': [{ name: 'file1.png', size: 123, contents: '' }, { name: 'file2.png', size: 456, contents: '' }],
};
const saveTemplateData = {
  id: 1,
  email_type: 'assign',
  name: formData['template-name'],
  email_subject: formData[EMAIL_TEMPLATE_SUBJECT_KEY],
  email_greeting: formData['email-template-greeting'],
  email_closing: formData['email-template-closing'],
  email_files: formData['email-template-files'],
};
const templateType = saveTemplateData.email_type;
const saveTemplateSpy = jest.spyOn(EcommerceApiService, 'saveTemplate');

const SaveTemplateButtonWrapper = props => (
  <MemoryRouter>
    <Provider store={store}>
      <SaveTemplateButton
        templateType={templateType}
        emailTemplateSource={EMAIL_TEMPLATE_SOURCE_NEW_EMAIL}
        setMode={() => {}}
        handleSubmit={submitFunction => () => submitFunction(formData)}
        {...props}
      />
    </Provider>
  </MemoryRouter>
);

describe('<SaveTemplateButton />', () => {
  it('renders correctly in disabled state', () => {
    const tree = renderer
      .create((
        <SaveTemplateButtonWrapper />
      ))
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders correctly in enabled state', () => {
    const tree = renderer
      .create((
        <SaveTemplateButtonWrapper disabled={false} />
      ))
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders correctly while saving a template', () => {
    const newStore = mockStore({
      emailTemplate: {
        saving: true,
      },
    });

    const tree = renderer
      .create((
        <SaveTemplateButtonWrapper store={newStore} />
      ))
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('calls saveTemplate on click with correct data', () => {
    const expectedActions = [{
      type: SAVE_TEMPLATE_REQUEST,
      payload: { emailType: templateType },
    }];
    const successResponse = {
      email_subject: saveTemplateData.email_subject,
      email_greeting: saveTemplateData.email_greeting,
      email_body: 'email body',
      email_closing: saveTemplateData.email_closing,
      email_files: saveTemplateData.email_files,
    };
    EcommerceApiService.saveTemplate.mockImplementation(() => Promise.resolve({
      data: successResponse,
    }));
    render((
      <SaveTemplateButtonWrapper disabled={false} />
    ));

    fireEvent.click(screen.getByText('Save Template'));
    expect(store.getActions()).toEqual(expectedActions);

    expect(saveTemplateSpy).toHaveBeenCalledWith(saveTemplateData);
  });

  // This is not possible in react-testing-library -> https://github.com/testing-library/react-testing-library/issues/624#issuecomment-605105191
  // Need to re-write this test in some better way.
  it.skip('saveTemplate raises correct errors on invalid data submission', () => {
    const invalidFormData = {
      'email-template-subject': 'G'.repeat(1001),
      'email-template-greeting': 'G'.repeat(50001),
      'email-template-closing': 'C'.repeat(50001),
    };
    const SaveTemplateButtonWrapperWithInvalidData = props => (
      <MemoryRouter>
        <Provider store={store}>
          <SaveTemplateButton
            templateType={templateType}
            setMode={() => {}}
            handleSubmit={submitFunction => () => submitFunction(invalidFormData)}
            {...props}
          />
        </Provider>
      </MemoryRouter>
    );
    render((
      <SaveTemplateButtonWrapperWithInvalidData disabled={false} />
    ));

    try {
      fireEvent.click(screen.getByText('Save Template'));
    } catch (e) {
      expect(e instanceof SubmissionError).toBeTruthy();
      expect(e.errors['template-name']).toEqual('No template name provided. Please enter a template name.');
      expect(e.errors['email-template-subject']).toEqual(`Email subject must be ${OFFER_ASSIGNMENT_EMAIL_SUBJECT_LIMIT} characters or less.`);
      expect(e.errors['email-template-greeting']).toEqual(`Email greeting must be ${EMAIL_TEMPLATE_FIELD_MAX_LIMIT} characters or less.`);
      expect(e.errors['email-template-closing']).toEqual(`Email closing must be ${EMAIL_TEMPLATE_FIELD_MAX_LIMIT} characters or less.`);
    }
  });
});
