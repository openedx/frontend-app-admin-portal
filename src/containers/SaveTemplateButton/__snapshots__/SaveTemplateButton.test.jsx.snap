// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SaveTemplateButton /> renders correctly in disabled state 1`] = `
<button
  aria-disabled={false}
  aria-live="assertive"
  className="pgn__stateful-btn pgn__stateful-btn-state-default save-template-btn btn btn-outline-primary"
  disabled={false}
  onClick={[Function]}
  type="button"
>
  <span
    className="d-flex align-items-center justify-content-center"
  >
    <span>
      Save Template
    </span>
  </span>
</button>
`;

exports[`<SaveTemplateButton /> renders correctly in enabled state 1`] = `
<button
  aria-disabled={false}
  aria-live="assertive"
  className="pgn__stateful-btn pgn__stateful-btn-state-default save-template-btn btn btn-outline-primary"
  disabled={false}
  onClick={[Function]}
  type="button"
>
  <span
    className="d-flex align-items-center justify-content-center"
  >
    <span>
      Save Template
    </span>
  </span>
</button>
`;

exports[`<SaveTemplateButton /> renders correctly while saving a template 1`] = `
<button
  aria-disabled={false}
  aria-live="assertive"
  className="pgn__stateful-btn pgn__stateful-btn-state-default save-template-btn btn btn-outline-primary"
  disabled={false}
  onClick={[Function]}
  type="button"
>
  <span
    className="d-flex align-items-center justify-content-center"
  >
    <span>
      Save Template
    </span>
  </span>
</button>
`;
